### ISSUE TITLE
WPS440/441 on multiple loops

### ISSUE THREAD

**Issue by Dreamsorcerer** [wemake-services/wemake-python-styleguide#1416]
With code like:
```
        for widget in self._widgets:
            await widget.update()
        ...
        for widget in self._widgets:
            await widget.set_something(state)
```

I get WPS440 and WPS441 due to reusing the widget variable name in the second loop.

I think that WPS441 shouldn't apply, as we are defining a new variable, as opposed to reusing the previous value from the last loop.

Then I think also, that WPS440 shouldn't apply, as the WPS441 stops you from reusing the first variable anyway, so shadowing a variable you're not supposed to be using is not an issue. i.e. If WPS441 is going to make you treat them as block-scoped variables, then shadowing them shouldn't be taken into account as they shouldn't be treated as being in the enclosing scope.

**Comment by anatoly-scherbakov** [comment]
Another example of the same issue with `except` blocks.

```python
        try:
            sqs_response = self.client.send_message(
                QueueUrl=self.url,
                MessageBody=message_body,
            )

        except self.client.exceptions.QueueDoesNotExist as err:
            raise SQSQueueDoesNotExist(queue=self) from err

        except self.client.exceptions.ClientError as err:
            if self._error_code_is(err, 'InvalidParameterValue'):
                raise MessageTooLarge(
                    max_supported_size=MAX_MESSAGE_SIZE,
                    message_body=message_body,
                )
```

cause 440 and 441:

```
  38:9     WPS440 Found block variables overlap: err
  except self.client.exceptions.ClientError as err:
  ^

  39:36    WPS441 Found control variable used after block: err
  if self._error_code_is(err, 'InvalidParameterValue'):
```

**Comment by huangyunzen** [comment]
Hello, @Elton-Lin and I would like to work on fixing this. Thanks!

**Comment by sobolevn** [comment]
Awesome! Thanks!

**Comment by huangyunzen** [comment]
> Another example of the same issue with `except` blocks.
> 
> ```python
>         try:
>             sqs_response = self.client.send_message(
>                 QueueUrl=self.url,
>                 MessageBody=message_body,
>             )
> 
>         except self.client.exceptions.QueueDoesNotExist as err:
>             raise SQSQueueDoesNotExist(queue=self) from err
> 
>         except self.client.exceptions.ClientError as err:
>             if self._error_code_is(err, 'InvalidParameterValue'):
>                 raise MessageTooLarge(
>                     max_supported_size=MAX_MESSAGE_SIZE,
>                     message_body=message_body,
>                 )
> ```
> 
> cause 440 and 441:
> 
> ```
>   38:9     WPS440 Found block variables overlap: err
>   except self.client.exceptions.ClientError as err:
>   ^
> 
>   39:36    WPS441 Found control variable used after block: err
>   if self._error_code_is(err, 'InvalidParameterValue'):
> ```

We believe that the issue with the except blocks is fixed in this issue:
https://github.com/wemake-services/wemake-python-styleguide/issues/1115?fbclid=IwAR0VijbU0sftzbXkWnkaeE0AFji5qR4Xlbx4LSKiiJreW4zoj58MPjnVua4

**Comment by huangyunzen** [comment]
@sobolevn Should we work on modifying both rules WPS 440/441 so that they don't trigger on multiple for loops?

**Comment by sobolevn** [comment]
Yes, @huangyunzen, please!

**Comment by huangyunzen** [comment]
WPS441 also applies to "with blocks" in addition to the for blocks mentioned in the issue - should we fix them too?

**Comment by sobolevn** [comment]
Yes, please 👍

**Comment by huangyunzen** [comment]
@sobolevn Is it possible to separate this issue into two issues? We are done with fixing WPS441, but WPS440 might take some more time and we were wondering if we can submit a pull request now for only 441.

**Comment by sobolevn** [comment]
I can always accept a PR with a partial fix! 👍

**Comment by sobolevn** [comment]
Done!

**Comment by fpuga** [comment]
Can this ticket be reopened?

I'm still getting a `WPS440` in code like the shown in the first comment:

```python
for widget in self._widgets:
    await widget.update()
...
for widget in self._widgets:
    await widget.set_something(state)
```

**Comment by sobolevn** [comment]
Reproduction:

```python
async def some():
    for widget in self._widgets:
        await widget.update()

    for widget in self._widgets:
        await widget.set_something(state)

```

<img width="752" alt="Снимок экрана 2021-06-24 в 19 14 09" src="https://user-images.githubusercontent.com/4660275/123297592-5ebc4a80-d520-11eb-8a10-bc2e121e6500.png">

**Comment by strmwalker** [comment]
Any updates?

**Comment by Xoriun** [comment]
if the second loop if a comprehension, WPS441 is raised in the comprehension variable:

```
   for widget in self._widgets:
      await widget.update()
   ...
   return [widget.get_value() for widget in self._widgets]
```

**Comment by Xoriun** [comment]
```
cond = False
index = 4

if cond:
    for index in range(465):
        print(index)
else:
    print(index)
```

This raises an `WPS440` error in the last row, even though if execution reached that point, the loop would not have been executed and `index` would still be `4`.

**Comment by sobolevn** [comment]
`WPS440` is removed. This code no longer produces any violations.
